primitives/state-machine/src/trie_backend_essence.rs

```diff
@@ -429,246 +588,41 @@ where
-		prefix: Option<&[u8]>,
-		start_at: Option<&[u8]>,
-		f: F,
-	) {
-		let root = if let Some(child_info) = child_info.as_ref() {
-			match self.child_root(child_info) {
-				Ok(Some(v)) => v,
-				// If the child trie doesn't exist, there is no need to continue.
-				Ok(None) => return,
-				Err(e) => {
-					debug!(target: "trie", "Error while iterating child storage: {}", e);
-					return
-				},
-			}
+	pub fn raw_iter(&self, args: IterArgs) -> Result<RawIter<S, H, C>> {
+		let root = if let Some(child_info) = args.child_info.as_ref() {
+			let root = match self.child_root(&child_info)? {
+				Some(root) => root,
+				None => return Ok(Default::default()),
+			};
+			root
 		} else {
 			self.root
 		};
 
-		self.trie_iter_key_inner(&root, prefix, f, child_info, start_at)
-	}
-
-	/// Execute given closure for all keys starting with prefix.
-	pub fn for_child_keys_with_prefix(
-		&self,
-		child_info: &ChildInfo,
-		prefix: &[u8],
-		mut f: impl FnMut(&[u8]),
-	) {
-		let root = match self.child_root(child_info) {
-			Ok(Some(v)) => v,
-			// If the child trie doesn't exist, there is no need to continue.
-			Ok(None) => return,
-			Err(e) => {
-				debug!(target: "trie", "Error while iterating child storage: {}", e);
-				return
-			},
-		};
-
-		self.trie_iter_key_inner(
-			&root,
-			Some(prefix),
-			|k| {
-				f(k);
-				true
-			},
-			Some(child_info),
-			None,
-		)
-	}
-
-	/// Execute given closure for all keys starting with prefix.
-	pub fn for_keys_with_prefix<F: FnMut(&[u8])>(&self, prefix: &[u8], mut f: F) {
-		self.trie_iter_key_inner(
-			&self.root,
-			Some(prefix),
-			|k| {
-				f(k);
-				true
-			},
-			None,
-			None,
-		)
-	}
-
-	fn trie_iter_key_inner<F: FnMut(&[u8]) -> bool>(
-		&self,
-		root: &H::Out,
-		maybe_prefix: Option<&[u8]>,
-		mut f: F,
-		child_info: Option<&ChildInfo>,
-		maybe_start_at: Option<&[u8]>,
-	) {
-		let mut iter = move |db| -> sp_std::result::Result<(), Box<TrieError<H::Out>>> {
-			self.with_recorder_and_cache(Some(*root), |recorder, cache| {
-				let trie = TrieDBBuilder::<H>::new(db, root)
-					.with_optional_recorder(recorder)
-					.with_optional_cache(cache)
-					.build();
-				let prefix = maybe_prefix.unwrap_or(&[]);
-				let iter = match maybe_start_at {
-					Some(start_at) =>
-						TrieDBKeyIterator::new_prefixed_then_seek(&trie, prefix, start_at),
-					None => TrieDBKeyIterator::new_prefixed(&trie, prefix),
-				}?;
-
-				for x in iter {
-					let key = x?;
-
-					debug_assert!(maybe_prefix
-						.as_ref()
-						.map(|prefix| key.starts_with(prefix))
-						.unwrap_or(true));
-
-					if !f(&key) {
-						break
-					}
-				}
-
-				Ok(())
-			})
-		};
-
-		let result = if let Some(child_info) = child_info {
-			let db = KeySpacedDB::new(self, child_info.keyspace());
-			iter(&db)
-		} else {
-			iter(self)
-		};
-		if let Err(e) = result {
-			debug!(target: "trie", "Error while iterating by prefix: {}", e);
+		if self.root == Default::default() {
```

Comment:
It's always a bit awkward to tell if self.root is vec![0; 32] or hash(trie_empty_node) (eg MemoryDb::null_node_data), actually it looks like self.empty could be use instead (that is the root of a empty trie).

Reply:
Hmmm... why do we have `self.empty` anyway? In practice won't `H::default` always point to an empty storage root and a value which doesn't exist? (And idiomatically `T::default()` is essentially always used to denote an empty something, e.g. `Vec::default`, `HashMap::default`, etc. are all empty, so it's kinda weird to have two things which mean "empty".)

But anyhow, as far as I can see this changing this to `self.root == self.empty` will fail at least one test (`pairs_are_empty_on_empty_storage`) because the `root` there is initialized with `Default::default`, so the fake empty key `self.empty` won't match, so it will go and call `TrieDBRawIterator::new_prefixed`, and that will fail with an `Invalid state root` error. (This worked previously without special-casing it like this because `pairs` just ignored the errors, and this new iterator doesn't do that anymore.)

Reply:
H::default is [0; 32], that's an invalid trie root. Could have given the actual empty trie value but that's not convenient at all (makes the H depends on trie internals). In practice it 's just one step from saying any invalid root is an empty trie, but that would be confusing (like if you got a root and no content you cannot say if it is empty).
But yes that's rather awkward that some place allows H::default.